dlrover/python/master/diagnosis/diagnosis_manager.py

```diff
@@ -53,103 +53,170 @@
 from dlrover.python.master.diagnosis.diagnosis_data_manager import (
     DiagnosisDataManager,
 )
-from dlrover.python.master.diagnosis.precheck_operator import (
-    NoPreCheckOperator,
-)
 from dlrover.python.master.node.job_context import get_job_context
+from dlrover.python.scheduler.job import JobArgs
+from dlrover.python.util.function_util import TimeoutException, timeout
+from dlrover.python.util.time_util import get_pending_timeout
 
 _metric_context = JobMetricContext.singleton_instance()
 _dlrover_context = Context.singleton_instance()
 
 
+# pending timeout + 600s
+def get_pre_check_timeout():
+    return get_pending_timeout() + 600
+
+
 class DiagnosisManager:
     """
-    DiagnosisManager is to manage all diagnosis issues in a training job
-
+    DiagnosisManager is used to manage all diagnosis issues in a training job.
     """
 
-    def __init__(self, job_name=None):
+    def __init__(self, job_args: JobArgs = None):
         self._is_observing_started = False
         self._data_manager: DiagnosisDataManager = DiagnosisDataManager(600)
         self._diagnostician: Diagnostician = Diagnostician(self._data_manager)
         self._job_context = get_job_context()
-        self._job_name = job_name
+        self._job_args = job_args
         self._metric_monitor = None
         self._lock = threading.Lock()
 
-    @classmethod
-    def get_pre_check_operators(cls):
-        return [NoPreCheckOperator()]
-
     def collect_diagnosis_data(self, data: DiagnosisData):
         self._data_manager.store_data(data)
 
+    @timeout(callback_func=get_pre_check_timeout)
     def pre_check(self):
-        if not _dlrover_context.pre_check_enabled:
+        if not _dlrover_context.pre_check_enabled():
+            logger.info(
+                "Pre-check operator config is empty, " "pre-check disabled."
+            )
             return
 
         start = time.time()
-        pre_check_ops = self.get_pre_check_operators()
+        job_ctx = get_job_context()
+        pre_check_ops = _dlrover_context.get_pre_check_operators()
         logger.info(
-            "Start to training pre-check with "
-            f"operators: {[op.__class__.__name__ for op in pre_check_ops]}."
+            "Start training pre-check with "
+            f"operators: {[op.__class__.__name__ for op in pre_check_ops]} "
+            f"under timeout: {get_pre_check_timeout()}s."
         )
 
-        for pre_check_op in pre_check_ops:
-            current_start = time.time()
-            current_op_result = None
-            pre_check_op_name = pre_check_op.__class__.__name__
+        round = 0
+        pre_check_finish = False
+
+        while True:
```

Comment:
1. Add comments to explain the key idea here. For example, when an operator fails, it will bypass/retry/execute failure action. 
2. When one operator fails, why retry all precheck operators ?

Reply:
1. Done.
2. Yes. Here's a simple example: if the scheduling check is performed first, followed by the node check, and it is found that some nodes need to be replaced, then after triggering the relaunch logic, the scheduling check needs to be performed again to ensure that the relaunched nodes can be scheduled properly.